# Pipeline test with K-best XGB feature selection
The notebook focuses on the testing of K-best XGB feature selection with a variety parameter combinations. 

The classifiers that will be tested are:
1) XGBoost
2) SVM
3) Logistic Regression

In [18]:
# Whats needed....
from Supervised_ML_Pipeline import MLPipeline, add_inne_features
import pandas as pd

In [19]:
# Checking for available models
MLPipeline.list_models()

['logistic_regression', 'xgboost', 'svm']

In [20]:
# Checking for available undersampling methods
MLPipeline.list_undersample_methods()

['RandomUnderSampler', 'CLEANSE']

In [21]:
# Checking for available undersampling methods
MLPipeline.list_feature_selection_methods()

['k_best_catboost', 'k_best_xgb', 'k_best_lgbm', 'k_best_decision_tree']

In [22]:
# Setting up The dataset
df = pd.read_csv("/dsa/groups/casestudycf25/team02/silver/unified_dataset.csv")



exclude_cols = [
    'npi', 'rfrg_prvdr_state_abrvtn',
    'specialty_type', 'specialty'
]

df.head(5)

,npi,rfrg_prvdr_state_abrvtn,year,target,avg_suplr_mdcr_pymt_amt_sum,tot_suplr_mdcr_pymt_amt_hcpcs_rentl_ind_z_scr_sum,tot_suplr_mdcr_pymt_amt_hcpcs_rentl_ind_z_scr_median,avg_suplr_mdcr_alowd_amt_hcpcs_rentl_ind_rat_median,avg_suplr_mdcr_pymt_amt_min,tot_suplr_mdcr_pymt_amt_hcpcs_rentl_ind_z_scr_min,...,claims_per_bene_zscore_by_type,claims_per_bene_outlier_by_type,services_per_bene_outlier,services_per_bene_zscore_by_type,services_per_bene_outlier_by_type,benes_per_supplier_outlier,benes_per_supplier_zscore_by_type,benes_per_supplier_outlier_by_type,elderly_patient_concentration,young_patient_concentration
0,1003000597,OK,2022,0,1.692306,0.166923,0.166923,1.051990,1.692306,0.166923,...,3.003923,True,True,4.288024,True,False,-0.187419,False,0.0,0.0
1,1003000597,OK,2023,0,23.833223,-0.150274,-0.226470,1.001578,0.112054,-0.475409,...,1.564890,False,True,2.626744,True,False,-0.159271,False,0.0,0.0
2,1003000902,KY,2021,0,155.711788,-2.380244,-0.336014,1.000711,0.087436,-0.660212,...,-0.036876,False,False,-0.073170,False,False,-0.092180,False,0.0,0.0
3,1003000902,KY,2022,0,67.361088,-2.190131,-0.381036,1.000020,0.028794,-0.621005,...,-0.374303,False,False,0.108292,False,False,-0.098162,False,0.0,0.0
4,1003000902,KY,2023,0,40.802223,-1.466342,-0.355328,1.004687,0.098643,-0.463672,...,-0.049636,False,False,0.152416,False,False,-0.091872,False,0.0,0.0


## Attempt - 1 (XGBoost with Default K-best XGB)

Scikit-learns default k value is 10.

In [25]:
# Grid 1: Tree structure with learning rate
param_grid_1 = [
    {
        "model__max_depth": [3, 5],
        "model__min_child_weight": [1, 3],
        "model__learning_rate": [0.05, 0.1],
        "model__n_estimators": [100, 200]
    }
]

# Grid 2: Sampling with regularization
param_grid_2 = [
    {
        "model__subsample": [0.6, 1.0],
        "model__colsample_bytree": [0.6, 1.0],
        "model__reg_alpha": [0, 0.1],
        "model__reg_lambda": [1, 5]
    }
]

# Grid 3: Imbalance handling
param_grid_3 = [
    {
        "model__scale_pos_weight": [50, 100, 200],
        "model__max_delta_step": [1, 5],
        "model__min_child_weight": [1, 5],
        "model__learning_rate": [0.01, 0.05],
        "model__n_estimators": [200, 300]
    }
]

The above parameters determine how well the model will perform. Based on the performance, these parameters will be adjusted.

In [26]:
pipe = MLPipeline(model_name="xgboost", target_col="target")

In [27]:
# param_grid_1
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe.train(df, param_grid=param_grid_1, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 16 candidates, totalling 80 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2122 ± 0.1749 (std across folds)
Score range: min = 0.0763  max = 0.5524  cv = 0.8245 (std/mean)
Training time: 2.1 Seconds
Best params: {'model__learning_rate': 0.1, 'model__max_depth': 5, 'model__min_child_weight': 1, 'model__n_estimators': 100}

Per-fold breakdown:
  Fold          Score
  1            0.5524
  2            0.2011
  3            0.1106
  4            0.0763
  5            0.1206


------------------------------------------------------------------------------------------------------------------------------------------------------



Undersampl

In [28]:
# param_grid_2
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe.train(df, param_grid=param_grid_2, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 16 candidates, totalling 80 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2024 ± 0.1589 (std across folds)
Score range: min = 0.0813  max = 0.5111  cv = 0.7850 (std/mean)
Training time: 2.0 Seconds
Best params: {'model__colsample_bytree': 1.0, 'model__reg_alpha': 0, 'model__reg_lambda': 1, 'model__subsample': 0.6}

Per-fold breakdown:
  Fold          Score
  1            0.5111
  2            0.1920
  3            0.1286
  4            0.0813
  5            0.0989


------------------------------------------------------------------------------------------------------------------------------------------------------



Undersampling (C

In [29]:
# param_grid_3
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe.train(df, param_grid=param_grid_3, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 48 candidates, totalling 240 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2399 ± 0.1860 (std across folds)
Score range: min = 0.0938  max = 0.5886  cv = 0.7750 (std/mean)
Training time: 7.3 Seconds
Best params: {'model__learning_rate': 0.01, 'model__max_delta_step': 5, 'model__min_child_weight': 1, 'model__n_estimators': 300, 'model__scale_pos_weight': 50}

Per-fold breakdown:
  Fold          Score
  1            0.5886
  2            0.1325
  3            0.2759
  4            0.0938
  5            0.1089


------------------------------------------------------------------------------------------------------------------------------

## Obervations:
With the K value begin set as 10, with a score of 0.2399, these parameters performed the best: 
1) model__learning_rate: 0.01
2) model__max_delta_step: 5
3) model__min_child_weight: 1
4) model__n_estimators: 300
5) model__scale_pos_weight: 50
6) Undersampling : RandomUnderSampler

When using cleanse clustering, most of the performance also saw a significant dip. Further testing will have to be conducted to see how it performs. 

Maybe adjusting the number of K values for the feature selection copuld yield us some better results.

## Attempt 2 (XGBoost with K-best XGB with k values)

In [6]:
# Grid 1: Tree structure with learning rate
param_grid_1 = [
    {
        "selector__k": [10, 15, 20, 25],
        "model__max_depth": [3, 5, 7],
        "model__min_child_weight": [1, 3, 5],
        "model__learning_rate": [0.01, 0.05, 0.1],
        "model__n_estimators": [100, 200, 300]
    }
]

# Grid 2: Sampling with regularization
param_grid_2 = [
    {
        "selector__k": [10, 15, 20, 25],
        "model__subsample": [0.6, 0.8, 1.0],
        "model__colsample_bytree": [0.6, 0.8, 1.0],
        "model__reg_alpha": [0, 0.1, 1.0],
        "model__reg_lambda": [1, 5, 10]
    }
]

# Grid 3: Imbalance handling
param_grid_3 = [
    {
        "selector__k": [10, 15, 20, 25],
        "model__scale_pos_weight": [10, 50, 100, 200],
        "model__max_delta_step": [1, 5],
        "model__min_child_weight": [1, 5, 10],
        "model__learning_rate": [0.01, 0.05],
        "model__n_estimators": [200, 300]
    }
]

In [7]:
pipe2 = MLPipeline(model_name="xgboost", target_col="target")

In [8]:
# param_grid_1
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe2.train(df, param_grid=param_grid_1, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 324 candidates, totalling 1620 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2496 ± 0.1640 (std across folds)
Score range: min = 0.0716  max = 0.5088  cv = 0.6570 (std/mean)
Training time: 50.2 Seconds
Best params: {'model__learning_rate': 0.1, 'model__max_depth': 5, 'model__min_child_weight': 1, 'model__n_estimators': 200, 'selector__k': 20}

Per-fold breakdown:
  Fold          Score
  1            0.5088
  2            0.0949
  3            0.2193
  4            0.0716
  5            0.3534


---------------------------------------------------------------------------------------------------------------------------------------------

In [9]:
# param_grid_2
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe2.train(df, param_grid=param_grid_2, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 324 candidates, totalling 1620 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2608 ± 0.1586 (std across folds)
Score range: min = 0.0870  max = 0.5090  cv = 0.6082 (std/mean)
Training time: 31.0 Seconds
Best params: {'model__colsample_bytree': 1.0, 'model__reg_alpha': 0.1, 'model__reg_lambda': 10, 'model__subsample': 1.0, 'selector__k': 20}

Per-fold breakdown:
  Fold          Score
  1            0.5090
  2            0.1133
  3            0.2281
  4            0.0870
  5            0.3666


------------------------------------------------------------------------------------------------------------------------------------------------

In [10]:
# param_grid_3
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe2.train(df, param_grid=param_grid_3, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 384 candidates, totalling 1920 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2663 ± 0.1713 (std across folds)
Score range: min = 0.1024  max = 0.5606  cv = 0.6432 (std/mean)
Training time: 62.3 Seconds
Best params: {'model__learning_rate': 0.01, 'model__max_delta_step': 5, 'model__min_child_weight': 5, 'model__n_estimators': 300, 'model__scale_pos_weight': 10, 'selector__k': 15}

Per-fold breakdown:
  Fold          Score
  1            0.5606
  2            0.1032
  3            0.2273
  4            0.1024
  5            0.3381


--------------------------------------------------------------------------------------------------------

## Observation:

With K set at 15, the a score of 0.2663 was acquired with this parameters:
1) model__learning_rate: 0.01
2) model__max_delta_step: 5
3) model__min_child_weight: 5
4) model__n_estimators: 300
5) model__scale_pos_weight: 10
6) Undersampling : RandomUnderSampler

The increase in performance is quite small but still significant. 

Interestingly the performance of the model using Cleanse did inprove alot with more features to work with.
    
Score - 0.0295
1) model__colsample_bytree: 0.8
2) model__reg_alpha: 0.1
3) model__reg_lambda: 1
4) model__subsample: 0.6
5) Undersampling : CLEANSE

Maybe a lilttle bit more tweakinhg can fetch us better results.

In [14]:

param_grid_2 = [
    {
        "selector__k": [15, 20, 25, 50, 100, 200],
        "model__subsample": [0.6, 0.8, 1.0],
        "model__colsample_bytree": [0.6, 0.8, 1.0],
        "model__reg_alpha": [0, 0.1, 1.0],
        "model__reg_lambda": [0.5, 1, 5, 10]
    }
]

param_grid_3 = [
    {
        "selector__k": [20, 25, 50, 100, 200],
        "model__scale_pos_weight": [10, 50, 100, 200],
        "model__max_delta_step": [1, 5, 10, 15],
        "model__min_child_weight": [1, 5, 10],
        "model__learning_rate": [0.005, 0.01, 0.05],
        "model__n_estimators": [300, 350, 400]
    }
]

In [16]:
pipe2 = MLPipeline(model_name="xgboost", target_col="target")

In [26]:
# param_grid_3
pipe2.train(df, param_grid=param_grid_3, group_col="npi", exclude_cols=exclude_cols, undersample_method='RandomUnderSampler', feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 2160 candidates, totalling 10800 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2726 ± 0.1631 (std across folds)
Score range: min = 0.0888  max = 0.5170  cv = 0.5985 (std/mean)
Training time: 971.5 Seconds
Best params: {'model__learning_rate': 0.05, 'model__max_delta_step': 5, 'model__min_child_weight': 5, 'model__n_estimators': 350, 'model__scale_pos_weight': 50, 'selector__k': 20}

Per-fold breakdown:
  Fold          Score
  1            0.5170
  2            0.0888
  3            0.3120
  4            0.0941
  5            0.3508


-----------------------------------------------------------------------------------------------------

In [28]:
# param_grid_2
pipe2.train(df, param_grid=param_grid_2, group_col="npi", exclude_cols=exclude_cols, undersample_method='CLEANSE', feature_selection_method='k_best_xgb')


Undersampling (CLEANSE) applied.
Class distribution after undersampling: {0: 67659, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 648 candidates, totalling 3240 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.0295 ± 0.0573 (std across folds)
Score range: min = 0.0004  max = 0.1440  cv = 1.9419 (std/mean)
Training time: 2056.8 Seconds
Best params: {'model__colsample_bytree': 0.8, 'model__reg_alpha': 0.1, 'model__reg_lambda': 1, 'model__subsample': 0.6, 'selector__k': 20}

Per-fold breakdown:
  Fold          Score
  1            0.1440
  2            0.0004
  3            0.0004
  4            0.0009
  5            0.0018


------------------------------------------------------------------------------------------------------------------------------------------------------



## Obervations:

The performance increase is quite small.

What if we ran INNE first ?

In [23]:
df1 = add_inne_features(
    df,

    # Default INNE parameters that were found to be best during standalone tuning.
    # You can adjust these if you want to test different INNE configurations.
    transaction_data_path="/dsa/groups/casestudycf25/team02/silver/cms_general_payments_anomaly_ready.csv",
    group_col="npi",
    year_col="year",
    txn_provider_col="covered_recipient_npi",
    txn_year_col="program_year",
    txn_target_col="target",
    txn_id_cols=["covered_recipient_npi", "record_id", "program_year"],
    n_estimators=100,
    max_samples=8,
    contamination=0.0005,
    top_k_features=15,
    null_fill=-1,
    verbose=1,
)

INNE: Loading /dsa/groups/casestudycf25/team02/silver/cms_general_payments_anomaly_ready.csv
INNE: 932908 transactions, 15 features selected
INNE: Fitted with n_estimators=100, max_samples=8, contamination=0.0005
INNE: Score range=[0.1916, 1.0000], P95 cutoff=0.9151
INNE: Rolled up to 158751 provider-years, 18285 with any above P95
INNE: Joined -> (140827, 219), 39586/140827 matched (28.1%)


In [24]:
## param_grid_2
pipe2 = MLPipeline(model_name="xgboost", target_col="target")
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe2.train(df1, param_grid=param_grid_2, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 648 candidates, totalling 3240 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2566 ± 0.1659 (std across folds)
Score range: min = 0.0706  max = 0.5238  cv = 0.6465 (std/mean)
Training time: 477.1 Seconds
Best params: {'model__colsample_bytree': 1.0, 'model__reg_alpha': 0, 'model__reg_lambda': 1, 'model__subsample': 1.0, 'selector__k': 20}

Per-fold breakdown:
  Fold          Score
  1            0.5238
  2            0.0940
  3            0.3274
  4            0.0706
  5            0.2670


--------------------------------------------------------------------------------------------------------------------------------------------------

In [ ]:
## param_grid_3
pipe2 = MLPipeline(model_name="xgboost", target_col="target")
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe2.train(df1, param_grid=param_grid_3, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 2160 candidates, totalling 10800 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: xgboost
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2648 ± 0.1859 (std across folds)
Score range: min = 0.1033  max = 0.5995  cv = 0.7018 (std/mean)
Training time: 2663.1 Seconds
Best params: {'model__learning_rate': 0.005, 'model__max_delta_step': 10, 'model__min_child_weight': 1, 'model__n_estimators': 300, 'model__scale_pos_weight': 200, 'selector__k': 25}

Per-fold breakdown:
  Fold          Score
  1            0.1191
  2            0.5995
  3            0.3318
  4            0.1705
  5            0.1033


-------------------------------------------------------------------------------------------------

## Attempt - 3 (SVM with K-best XGB with k values)

Due to the nature of CLEANSE undersampling, it will not be used here. 

In [15]:
# SVM Grid 1: Kernel and regularization
param_grid_svm_1 = [
    {
        "selector__k": [10, 15, 20, 25],
        "model__C": [0.1, 1, 10, 100],
        "model__kernel": ["rbf", "linear"],
        "model__gamma": ["scale", "auto"],
    }
]

# SVM Grid 2: Class imbalance handling
param_grid_svm_2 = [
    {
        "selector__k": [10, 15, 20, 25],
        "model__C": [1, 10, 100],
        "model__class_weight": ["balanced", None],
        "model__kernel": ["rbf"],
        "model__gamma": ["scale", "auto"],
    }
]

In [16]:
pipe3 = MLPipeline(model_name="svm", target_col="target")

In [17]:
# param_grid_svm_1
pipe3.train(df, param_grid=param_grid_svm_1, group_col="npi", exclude_cols=exclude_cols, undersample_method='RandomUnderSampler', feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 64 candidates, totalling 320 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: svm
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2634 ± 0.1681 (std across folds)
Score range: min = 0.0810  max = 0.4733  cv = 0.6381 (std/mean)
Training time: 13.4 Seconds
Best params: {'model__C': 100, 'model__gamma': 'scale', 'model__kernel': 'rbf', 'selector__k': 25}

Per-fold breakdown:
  Fold          Score
  1            0.4733
  2            0.1819
  3            0.4574
  4            0.0810
  5            0.1233


------------------------------------------------------------------------------------------------------------------------------------------------------




In [18]:
# param_grid_svm_2
pipe3.train(df, param_grid=param_grid_svm_2, group_col="npi", exclude_cols=exclude_cols, undersample_method='RandomUnderSampler', feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 48 candidates, totalling 240 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: svm
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.3029 ± 0.2037 (std across folds)
Score range: min = 0.1203  max = 0.6419  cv = 0.6727 (std/mean)
Training time: 4.5 Seconds
Best params: {'model__C': 100, 'model__class_weight': 'balanced', 'model__gamma': 'scale', 'model__kernel': 'rbf', 'selector__k': 10}

Per-fold breakdown:
  Fold          Score
  1            0.6419
  2            0.2017
  3            0.4295
  4            0.1211
  5            0.1203


------------------------------------------------------------------------------------------------------------------------------------------------------




## Obervation:

Interestingly, SVM performs really well on just 10 features. 

Score - 0.3029

1) model__C : 100
2) model__class_weight : balanced
3) model__gamma : scale
4) model__kernel : rbf
5) selector__k: 10

We can try using poly or sigmoid kernal, which can build out more complex boundries, possibly catching the complex relations and in turn improving performance. 

## Attempt - 4

In [17]:
param_grid_svm_1 = [
    # RBF kernel
    {
        "selector__k": [5, 10, 15, 20, 25],
        "model__C": [1, 10, 100, 150, 200],
        "model__class_weight": ["balanced", None, {0: 1, 1: 10}, {0: 1, 1: 50}, {0: 1, 1: 100}],
        "model__kernel": ["rbf"],
        "model__gamma": ["scale", "auto", 0.001, 0.01, 0.1]
    },
    # Sigmoid kernel
    {
        "selector__k": [5, 10, 15, 20, 25],
        "model__C": [1, 10, 100, 150, 200],
        "model__class_weight": ["balanced", None, {0: 1, 1: 10}, {0: 1, 1: 50}, {0: 1, 1: 100}],
        "model__kernel": ["sigmoid"],
        "model__gamma": ["scale", "auto", 0.001, 0.01, 0.1],
        "model__coef0": [0.0, 0.1, 0.5]
    },
    # Poly kernel
    {
        "selector__k": [5, 10, 15, 20, 25],
        "model__C": [1, 10, 100, 150, 200],
        "model__class_weight": ["balanced", None, {0: 1, 1: 10}, {0: 1, 1: 50}, {0: 1, 1: 100}],
        "model__kernel": ["poly"],
        "model__gamma": ["scale", "auto", 0.001, 0.01, 0.1],
        "model__degree": [2, 3, 4],
        "model__coef0": [0.0, 0.1, 0.5]
    },
]

In [18]:
pipe3 = MLPipeline(model_name="svm", target_col="target")

In [19]:
# param_grid_svm_1
pipe3.train(df, param_grid=param_grid_svm_1, group_col="npi", exclude_cols=exclude_cols, undersample_method='RandomUnderSampler', feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 8125 candidates, totalling 40625 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: svm
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.3365 ± 0.1917 (std across folds)
Score range: min = 0.1351  max = 0.6050  cv = 0.5698 (std/mean)
Training time: 1403.4 Seconds
Best params: {'model__C': 100, 'model__class_weight': {0: 1, 1: 50}, 'model__coef0': 0.0, 'model__degree': 4, 'model__gamma': 0.01, 'model__kernel': 'poly', 'selector__k': 25}

Per-fold breakdown:
  Fold          Score
  1            0.6050
  2            0.1351
  3            0.5271
  4            0.1794
  5            0.2358


------------------------------------------------------------------------------------------------------------

## Observation:

Ok model performs far better than its previous interations....

May adjusting some of the metrics could give a improved perfromance. 

In [22]:
param_grid_svm_1 = [
   # Sigmoid kernel
    {
        "selector__k": [5, 10, 15, 20, 25,50, 100, 150, 200],
        "model__C": [10, 100, 150],
        "model__class_weight": ["balanced", None, {0: 1, 1: 25}, {0: 1, 1: 50}, {0: 1, 1: 100}],
        "model__kernel": ["sigmoid"],
        "model__gamma": ["scale", "auto", 0.001, 0.01, 0.1],
        "model__coef0": [0.0, 0.01, 0.1, 0.5]
    },
    # Poly kernel
    {
        "selector__k": [5, 10, 15, 20, 25, 50, 100, 150, 200],
        "model__C": [ 10, 100, 150, 200],
        "model__class_weight": ["balanced", None, {0: 1, 1: 25}, {0: 1, 1: 50}, {0: 1, 1: 100}],
        "model__kernel": ["poly"],
        "model__gamma": ["scale", "auto", 0.001, 0.01, 0.1],
        "model__degree": [2, 3, 4, 6, 8],
        "model__coef0": [0.0, 0.01, 0.1, 0.5]
    },
]

In [23]:
# param_grid_svm_1
pipe3.train(df, param_grid=param_grid_svm_1, group_col="npi", exclude_cols=exclude_cols, undersample_method='RandomUnderSampler', feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 20700 candidates, totalling 103500 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: svm
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.3365 ± 0.1917 (std across folds)
Score range: min = 0.1351  max = 0.6050  cv = 0.5698 (std/mean)
Training time: 3510.8 Seconds
Best params: {'model__C': 100, 'model__class_weight': {0: 1, 1: 50}, 'model__coef0': 0.0, 'model__degree': 4, 'model__gamma': 0.01, 'model__kernel': 'poly', 'selector__k': 25}

Per-fold breakdown:
  Fold          Score
  1            0.6050
  2            0.1351
  3            0.5271
  4            0.1794
  5            0.2358


----------------------------------------------------------------------------------------------------------

## Observations:

The above results is the best so far. And as such is definitely a very good condidate.

What if we ran INNE first ?

In [ ]:
df1 = add_inne_features(
    df,

    # Default INNE parameters that were found to be best during standalone tuning.
    # You can adjust these if you want to test different INNE configurations.
    transaction_data_path="/dsa/groups/casestudycf25/team02/silver/cms_general_payments_anomaly_ready.csv",
    group_col="npi",
    year_col="year",
    txn_provider_col="covered_recipient_npi",
    txn_year_col="program_year",
    txn_target_col="target",
    txn_id_cols=["covered_recipient_npi", "record_id", "program_year"],
    n_estimators=100,
    max_samples=8,
    contamination=0.0005,
    top_k_features=15,
    null_fill=-1,
    verbose=1,
)

In [ ]:
param_grid_svm_1 = [
   # Sigmoid kernel
    {
        "selector__k": [5, 10, 15, 20, 25,50, 100, 150, 200],
        "model__C": [10, 100, 150],
        "model__class_weight": ["balanced", None, {0: 1, 1: 25}, {0: 1, 1: 50}, {0: 1, 1: 100}],
        "model__kernel": ["sigmoid"],
        "model__gamma": ["scale", "auto", 0.001, 0.01, 0.1],
        "model__coef0": [0.0, 0.01, 0.1, 0.5]
    },
    # Poly kernel
    {
        "selector__k": [5, 10, 15, 20, 25, 50, 100, 150, 200],
        "model__C": [ 10, 100, 150, 200],
        "model__class_weight": ["balanced", None, {0: 1, 1: 25}, {0: 1, 1: 50}, {0: 1, 1: 100}],
        "model__kernel": ["poly"],
        "model__gamma": ["scale", "auto", 0.001, 0.01, 0.1],
        "model__degree": [2, 3, 4, 6, 8],
        "model__coef0": [0.0, 0.01, 0.1, 0.5]
    },
]

In [ ]:
pipe3 = MLPipeline(model_name="svm", target_col="target")

# param_grid_svm_1
pipe3.train(df1, param_grid=param_grid_svm_1, group_col="npi", exclude_cols=exclude_cols, undersample_method='RandomUnderSampler', feature_selection_method='k_best_xgb')

## Attempt - 5

In [7]:
param_grid_lr = [
    # Grid 1: L1 and L2 regularization
    {
        "selector__k": [10, 15, 20, 25],
        "model__C": [0.001, 0.01, 0.1, 1, 10],
        "model__penalty": ["l1", "l2"],
        "model__solver": ["liblinear", "saga"],
        "model__class_weight": [None, "balanced"],
    },
    # Grid 2: Elasticnet
    {
        "selector__k": [10, 15, 20, 25],
        "model__C": [0.001, 0.01, 0.1, 1, 10],
        "model__penalty": ["elasticnet"],
        "model__solver": ["saga"],
        "model__l1_ratio": [0.2, 0.5, 0.8],
        "model__class_weight": [None, "balanced"],
    },
]

In [8]:
pipe4 = MLPipeline(model_name="logistic_regression", target_col="target")

In [9]:
# param_grid_lr
list = MLPipeline.list_undersample_methods()
for l in list:
    pipe4.train(df, param_grid=param_grid_lr, group_col="npi", exclude_cols=exclude_cols, undersample_method=l, feature_selection_method='k_best_xgb')


Undersampling (RandomUnderSampler) applied.
Class distribution after undersampling: {0: 377, 1: 34}

Feature Selection (k_best_xgb) selected.
Fitting 5 folds for each of 280 candidates, totalling 1400 fits

------------------------------------------------------------------------------------------------------------------------------------------------------


Model: logistic_regression
CV folds: 5  |  Scoring: average_precision
Best CV score: 0.2022 ± 0.1297 (std across folds)
Score range: min = 0.0785  max = 0.4473  cv = 0.6413 (std/mean)
Training time: 36.1 Seconds
Best params: {'model__C': 10, 'model__class_weight': None, 'model__penalty': 'l1', 'model__solver': 'liblinear', 'selector__k': 25}

Per-fold breakdown:
  Fold          Score
  1            0.0785
  2            0.1505
  3            0.4473
  4            0.1250
  5            0.2095


--------------------------------------------------------------------------------------------------------------------------------------------

## Obervations: 

The results with RandomUndersampler is the lowest yet. However, the performance with CLEANSE is not bad.

Maybe expanding on the parameters could yield improved results.